### DISCUSSION TITLE
npx / npm exec can't run packages with missing peerdep?

### DISCUSSION THREAD

**Discussion by thw0rted** [npm/feedback#561] [CLI] [answered]
I was trying to help someone who wanted to put together an `npx` one-liner to automate a task, as reliably/compatibly as possible.  One thing that tripped me up was that the package they wanted to run a binary from has a peer dependency that wasn't globally cached on my system.  `npm exec ts-node` installed the package but failed at runtime with "Cannot find module 'typescript'".

I did see in the docs that I can pass multiple instances of the `--package` flag, so I think this could be fixed by manually specifying every peerdep to be explicitly installed, but it might be nice to have a flag that would automate that, to support this sort of "fire and forget" use case.

**Comment by thw0rted** [comment, marked as answer]
https://github.com/npm/rfcs/pull/43 is enormous, can't tell if maybe this will Just Work in newer versions?  Will keep an eye on it.

**Reply by ljharb** [reply]
Make sure you’re using the latest version of npm 7.

**Reply by thw0rted** [reply]
Ah! Yeah, I'm still on Node 14 (LTS).  I switched to `npm@7` manually for a while but went back to 6.x during an upgrade.  If this is already fixed, I'll just wait until LTS picks up 7.x, no problem.

**Reply by ljharb** [reply]
You don't need to wait, you're always expected to update to the latest version of npm available - you're not supposed to just stick with the one that ships with your node version. npm 7 works on node 10+.

**Reply by thw0rted** [reply]
Hah, I thought this would be a good idea, so I did `npm i -g npm` in an Administrator mode git-bash, and managed to bork my whole installation.  I think I fixed it with the Node installer "repair" button, might try upgrading again on Monday.

**Reply by thw0rted** [reply]
FWIW, `npm` updates on Windows are still a bit of a mess.  `npm install -g npm@latest` broke my install, I had to "repair" it, then I still couldn't update it.  The most common answer on SO about how to do this says to use `npm-windows-upgrade` but that project hasn't had a release for 2 years and the maintainer recently archived the repo.  I finally got it working by finding the `npm` binary in use (with `which npm`), going to that directory, and running `npm update npm` (no `-g`).

@ljharb I can start a separate discussion if you like, but it'd be nice if the Node team had one official "how to upgrade NPM after you use one of our Windows MSI installers" page or something with the officially-supported way to do an in-place NPM upgrade.

**Comment by Maxim-Mazurok** [comment]
Actually, I'm not sure if that's the case.

It worked for me even without TS installed, as far as I can tell:

```
maxim@DESKTOP-804TKRG:~$ npx -y ts-node
> console.log(1+2)
3
undefined
>
(To exit, press Ctrl+C again or Ctrl+D or type .exit)
>
maxim@DESKTOP-804TKRG:~$ node
Welcome to Node.js v16.7.0.
Type ".help" for more information.
> require("typescript");
Uncaught Error: Cannot find module 'typescript'
Require stack:
- <repl>
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
    at Function.Module._load (node:internal/modules/cjs/loader:778:27)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
    at require (node:internal/modules/cjs/helpers:94:18) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [ '<repl>' ]
}
>
```

I also did `npm rm typescript` and `npm rm -g typescript` before trying.

**Reply by thw0rted** [reply]
This makes sense given @ljharb 's comment above -- I'm using Node 14 / npm 6.x, while you have Node 16 / npm 7.x.  They fixed npm 7.x to install peerdeps by default.
